system/lib/libc/dynlink.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2021 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ *
+ * Emscripten-specific version dlopen and associated functions.  Some code is
+ * shared with musl's ldso/dynlink.c.
+ */
+
+#define _GNU_SOURCE
+#include <assert.h>
+#include <dlfcn.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <pthread.h>
+#include <emscripten/emscripten.h>
+
+//#define DYLINK_DEBUG
+
+extern void *_dlopen_js(const char *file, int mode, void* handle);
+extern void *_dlsym_js(void *handle, const char *symbol);
+extern void _emscripten_dlopen_js(const char *filename, int flags, void* handle, em_arg_callback_func onsuccess, em_arg_callback_func onerror);
+
+struct dso {
+  struct dso *next, *prev;
+
+  // For async mode
+  em_dlopen_callback onsuccess;
+  em_arg_callback_func onerror;
+  void* user_data;
+
+  // Flexible array; must be final element of struct
+  char name[];
+};
+
+static struct dso *head, *tail;
+static pthread_rwlock_t lock;
+
+__attribute__((__visibility__("hidden")))
+void __dl_vseterr(const char *, va_list);
+
+static void error(const char *fmt, ...) {
+  va_list ap;
+  va_start(ap, fmt);
+  __dl_vseterr(fmt, ap);
+  va_end(ap);
+}
+
+__attribute__((__visibility__("hidden")))
```

Comment:
why hidden visibility for these? Does this file implement some API defined by musl?

Reply:
This code is copied directly from musl.. perhaps I could mark it as such.

Reply:
I ended up removing these .. the functions are already modified from the original musl versions and the command at the top says we have shared code so I think that is sufficient.